pages/editor.js

```diff
@@ -23,51 +24,70 @@ import {
   COLORS,
   DEFAULT_CODE
 } from '../lib/constants'
+import { mappings, reverseMappings, replaceQuery } from '../lib/routing'
 import { getState, saveState } from '../lib/util'
 
+const mapper = new Morph()
+
+const removeQueryString = str => {
+  const qI = str.indexOf('?')
+  return str.slice(0, qI)
+}
+
 class Editor extends React.Component {
-  /* pathname, asPath, err, req, res */
-  static async getInitialProps({ asPath }) {
+  static async getInitialProps({ asPath, query }) {
+    const path = removeQueryString(asPath)
+    const queryParams = mapper.map(mappings, query)
+    const initialState = Object.keys(queryParams).length ? queryParams : null
     try {
       // TODO fix this hack
-      if (asPath.length > 30) {
-        const content = await api.getGist(asPath)
-        return { content }
+      if (path.length > 30) {
+        const content = await api.getGist(path)
+        return { content, initialState }
       }
     } catch (e) {
       console.log(e)
     }
-    return {}
+    return { initialState }
   }
 
   constructor(props) {
     super(props)
-    this.state = {
-      background: 'rgba(171, 184, 195, 1)',
-      theme: THEMES.seti.id,
-      language: DEFAULT_LANGUAGE,
-      dropShadow: true,
-      windowControls: true,
-      widthAdjustment: true,
-      paddingVertical: '48px',
-      paddingHorizontal: '32px',
-      uploading: false,
-      code: props.content
-    }
+    this.state = Object.assign(
+      {
+        background: 'rgba(171, 184, 195, 1)',
+        theme: THEMES.seti.id,
+        language: DEFAULT_LANGUAGE,
+        dropShadow: true,
+        windowControls: true,
+        widthAdjustment: true,
+        paddingVertical: '48px',
+        paddingHorizontal: '32px',
+        uploading: false,
+        code: props.content,
+        _initialState: this.props.initialState
+      },
+      this.props.initialState
+    )
 
     this.save = this.save.bind(this)
     this.upload = this.upload.bind(this)
     this.updateCode = this.updateCode.bind(this)
   }
 
   componentDidMount() {
-    const state = getState(localStorage)
-    if (state) {
-      this.setState(state)
+    // Load from localStorage instead of query params
+    if (!this.state._initialState) {
+      const state = getState(localStorage)
+      if (state) {
+        this.setState(state)
+      }
     }
   }
 
   componentDidUpdate() {
+    const mappedState = mapper.map(reverseMappings, this.state)
```

Comment:
This should probably be done inside of `replaceQuery` so all that the editor component needs to do is pass a state object to a function. Also maybe rename to `updateQueryString` or something to make its use clear?

Reply:
Or another way, editor probably shouldn't have to know about `mapper` or `reverseMappings`